src/components/ToolsBar/useSlideImageUpload.tsx

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2024 Nordeck IT + Consulting GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useCallback } from 'react';
+import { FileRejection, useDropzone } from 'react-dropzone';
+import {
+  Size,
+  WhiteboardSlideInstance,
+  calculateCentredPosition,
+  calculateFittedElementSize,
+  useWhiteboardSlideInstance,
+} from '../../state';
+import {
+  defaultAcceptedImageTypes,
+  useImageUpload as useImageUploadContext,
+} from '../ImageUpload';
+import { whiteboardHeight, whiteboardWidth } from '../Whiteboard';
+
+/**
+ * Hook that uploads images to the current slide.
+ * Uses react-dropzone {@link https://react-dropzone.js.org/}.
+ */
+export function useSlideImageUpload() {
+  const slide = useWhiteboardSlideInstance();
+  const imageUpload = useImageUploadContext();
+
+  const handleDrop = useCallback(
+    async (files: File[], rejectedFiles: FileRejection[]) => {
+      const results = await imageUpload.handleDrop(files, rejectedFiles);
+
+      results.forEach((result) => {
+        if (result.status === 'fulfilled') {
+          addImageToSlide(
+            slide,
+            result.value.mxc,
+            result.value.fileName,
+            result.value.size,
+          );
+        }
+      });
```

Comment:
It add elements one by one, so a undo/redo stack will be updated for each element, is it intentional? Or if multiple elements are uploaded then we should undo with one undo action?

Reply:
Good point :+1:   I will change it to a batch action.